### PR TITLE
Issue 151: Add To Review slash command

### PR DESCRIPTION
**Pull request by SamirBoulil**
- [x] Unfurl url
- [x] Send dedicated message when PRDetails cannot be fetched
- [x] Remove {}
- [x] Remove dispatched_error
- [x] Add missing tests

### CODE REVIEW

File: config/routes.yaml [hunk 1]
```diff
@@ -3,6 +3,11 @@ chat_slack_messages:
     controller: slub.infrastructure.chat.slack.slub_bot_action::executeAction
     methods: [POST]
 
+chat_slack_tr:
```
Comment (SamirBoulil) [inline config/routes.yaml:6]: test

**Review by samirapptesting — APPROVED** [review]

**Review by samirapptesting — APPROVED** [review]

**Review by samirapptesting — CHANGES_REQUESTED** [review]
test

**Review by samirapptesting — CHANGES_REQUESTED** [review]
test

**Review by samirapptesting — APPROVED** [review]

**Review by samirapptesting — APPROVED** [review]
aze

**Review by samirapptesting — APPROVED** [review]
azeazeazeaze

**Review by samirapptesting — CHANGES_REQUESTED** [review]
azeazeaze

**Review by samirapptesting — CHANGES_REQUESTED** [review]
azeazeaze

**Review by samirapptesting — CHANGES_REQUESTED** [review]
azeazeaze

**Review by samirapptesting — CHANGES_REQUESTED** [review]
azeazeaze

File: src/Slub/Application/Common/ChatClient.php [hunk 2]
```diff
@@ -10,5 +10,8 @@ interface ChatClient
 {
     public function replyInThread(MessageIdentifier $messageIdentifier, string $text): void;
     public function setReactionsToMessageWith(MessageIdentifier $messageIdentifier, array $reactionsToSet): void;
-    public function publishInChannel(ChannelIdentifier $channelIdentifier, string $text);
+    public function publishInChannel(ChannelIdentifier $channelIdentifier, string $text): void;
+    public function answerWithEphemeralMessage(string $url, string $text): void;
+    public function acknowledgeRequest(string $url): void;
```
Comment (SamirBoulil) [inline src/Slub/Application/Common/ChatClient.php:15]: to remove

File: src/Slub/Infrastructure/Chat/Slack/APIHelper.php [hunk 3]
```diff
@@ -11,22 +11,35 @@
...
+                    $response->getStatusCode(),
                     json_encode($contents)
                 )
             );
         }
 
         return $contents;
     }
+
+    public static function checkStatusCodeSuccess(ResponseInterface $response): void
```
Comment (SamirBoulil) [inline src/Slub/Infrastructure/Chat/Slack/APIHelper.php:32]: to test

File: src/Slub/Infrastructure/Chat/Slack/ChannelIdentifierHelper.php [hunk 4]
```diff
@@ -16,8 +18,10 @@ public static function from(string $workspace, string $channel): string
         return sprintf('%s%s%s', $workspace, self::SEPARATOR, $channel);
     }
 
+    // TODO: Consider introducing 2 methods for that like (::slackWorkspace ::slackChannelIdentifier)
     public static function split(string $channelIdentifier): array
     {
+        Assert::contains($channelIdentifier, self::SEPARATOR, 'Impossible to split channel identifier');
```
Comment (SamirBoulil) [inline src/Slub/Infrastructure/Chat/Slack/ChannelIdentifierHelper.php:24]: to test

File: src/Slub/Infrastructure/Chat/Slack/ProcessTRRequestListener.php [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+use Slub\Infrastructure\VCS\Github\Query\GithubAPIHelper;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\TerminateEvent;
+use Symfony\Component\Routing\RouterInterface;
+use Webmozart\Assert\Assert;
+
+/**
+ * @author    Samir Boulil <samir.boulil@gmail.com>
+ */
+class ProcessTRRequestListener
```
Comment (SamirBoulil) [inline src/Slub/Infrastructure/Chat/Slack/ProcessTRRequestListener.php:24]: Rename ProcessTRSubmissionListener

Comment (SamirBoulil) [reply src/Slub/Infrastructure/Chat/Slack/ProcessTRRequestListener.php:24]: Add E2E tests for:
- no url
- URL correctly formatted but invalid PR Link (can't fetch details)
- Nominal: PR Submitted

Comment (SamirBoulil) [reply src/Slub/Infrastructure/Chat/Slack/ProcessTRRequestListener.php:24]: Consider, Reorganizing infra/chat/slack/* with sub directories (oAuth, TR submission, common, client ?)

File: src/Slub/Infrastructure/Chat/Slack/SlubBot.php [hunk 6]
```diff
@@ -93,10 +101,11 @@ private function listensToNewPR(BotMan $bot): void
                 $messageIdentifier,
                 $PRInfo
             );
+//            $bot->reply($putToReviewUsageChangeWarning);
         };
-        $bot->hears('.*TR.*<https://github.com/(.*)/pull/(\d+).*>.*$', $createNewPr);
-        $bot->hears('.*review.*<https://github.com/(.*)/pull/(\d+).*>.*$', $createNewPr);
-        $bot->hears('.*PR.*<https://github.com/(.*)/pull/(\d+).*>.*$', $createNewPr);
+//        $bot->hears('.*TR.*<https://github.com/(.*)/pull/(\d+).*>.*$', $createNewPr);
```
Comment (SamirBoulil) [inline src/Slub/Infrastructure/Chat/Slack/SlubBot.php:106]: uncomment before merging

File: src/Slub/Infrastructure/VCS/Github/Query/GetPRInfo.php [hunk 7]
```diff
@@ -113,4 +116,11 @@ private function deletions(array $PRDetails): int
     {
         return $PRDetails['deletions'];
     }
+
+    private function repositoryIdentifier(PRIdentifier $PRIdentifier): string
+    {
+        $breakout = GithubAPIHelper::breakoutPRIdentifier($PRIdentifier);
```
Comment (SamirBoulil) [inline src/Slub/Infrastructure/VCS/Github/Query/GetPRInfo.php:122]: remove breakout and introduce 2 different methods

**Review by samirapptesting — APPROVED** [review]
azeaze

**Review by samirapptesting — COMMENTED** [review]
azeaze

**Review by samirapptesting — APPROVED** [review]
azeaze

**Review by samirapptesting — CHANGES_REQUESTED** [review]
azeaze

File: tests/Integration/Infrastructure/Chat/Slack/TR/ProcessTRAsyncTest.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        $PRToReview = current($PRS);
+        self::assertEquals('SamirBoulil/slub/153', $PRToReview->PRIdentifier()->stringValue());
+        self::assertEquals('published_message_identifier', $PRToReview->messageIdentifiers()[0]->stringValue());
+        self::assertEquals('team_123@channel_name', $PRToReview->channelIdentifiers()[0]->stringValue());
+        self::assertEquals(self::USER_ID, $PRToReview->authorIdentifier()->stringValue());
+    }
+
+    private function assertToReviewMessageHasBeenPublished()
+    {
+//        $this->chatClientSpy->assertPublishMessageWithBlocksInChannelContains('team_123@channel_name', 'https://github.com/SamirBoulil/slub/pull/153');
```
Comment (SamirBoulil) [inline tests/Integration/Infrastructure/Chat/Slack/TR/ProcessTRAsyncTest.php:88]: uncomment

**Review by samirapptesting — COMMENTED** [review]
azeaze

**Review by samirapptesting — APPROVED** [review]
azeaze

**Review by samirapptesting — CHANGES_REQUESTED** [review]
azeaze

**Review by samirapptesting — COMMENTED** [review]
aze

**Review by samirapptesting — APPROVED** [review]
azeaze

**Pull request merged**